# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/layers/ipc/SharedSurfacesParent.h
# Commit: 11451cccc88c
# Full Hash: 11451cccc88c077e8fe763d4231bbf1af5226061
# Author: Andrew Osmond <aosmond@mozilla.com>
# Date: 2023-12-12 04:37:49
# Regressor Bug: 1869346
# File Overlap Count: 5
# Description:
#   Bug 1869346 - Rework how canvas accesses shared surfaces in the compositor process. r=gfx-reviewers,lsalzman
#   
#   This patch adds plumbing to allow sharing ExternalImageId via
#   SurfaceDescriptor. This will be used in a future patch by WebGL and
#   WebGPU to avoid extra copies. It also refactors how CanvasTranslator
# ==============================================================================

diff -r 8043d7693dba -r 11451cccc88c gfx/layers/ipc/SharedSurfacesParent.h
--- a/gfx/layers/ipc/SharedSurfacesParent.h	Mon Dec 11 18:15:34 2023 +0000
+++ b/gfx/layers/ipc/SharedSurfacesParent.h	Mon Dec 11 18:22:21 2023 +0000
@@ -129,6 +129,33 @@
   MappingTracker mTracker;
 };
 
+/**
+ * Helper class that is used to keep SourceSurfaceSharedDataWrapper objects
+ * around as long as one of the dependent IPDL actors is still alive and may
+ * reference them for a given PCompositorManager namespace.
+ */
+class SharedSurfacesHolder final {
+  NS_INLINE_DECL_THREADSAFE_REFCOUNTING(SharedSurfacesHolder)
+
+ public:
+  explicit SharedSurfacesHolder(uint32_t aNamespace) : mNamespace(aNamespace) {}
+
+  already_AddRefed<gfx::DataSourceSurface> Get(const wr::ExternalImageId& aId) {
+    uint32_t extNamespace = static_cast<uint32_t>(wr::AsUint64(aId) >> 32);
+    if (NS_WARN_IF(extNamespace != mNamespace)) {
+      MOZ_ASSERT_UNREACHABLE("Wrong namespace?");
+      return nullptr;
+    }
+
+    return SharedSurfacesParent::Get(aId);
+  }
+
+ private:
+  ~SharedSurfacesHolder() { SharedSurfacesParent::RemoveAll(mNamespace); }
+
+  uint32_t mNamespace;
+};
+
 }  // namespace layers
 }  // namespace mozilla
 